sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/PartitionLoadBalancerImpl.java

```diff
@@ -60,64 +61,72 @@ public PartitionLoadBalancerImpl(
+            executorService.execute(thread);
         });
     }
 
     @Override
     public Mono<Void> stop() {
-        return Mono.fromRunnable( () -> {
-            synchronized (lock) {
-                this.started = false;
-                this.cancellationTokenSource.cancel();
-            }
+        synchronized (lock) {
+            this.started = false;
+            this.cancellationTokenSource.cancel();
+        }
 
-            this.partitionController.shutdown().block();
-            this.cancellationTokenSource = null;
-        });
+        return this.partitionController.shutdown();
     }
 
     private Mono<Void> run(CancellationToken cancellationToken) {
         PartitionLoadBalancerImpl self = this;
 
-        return Mono.fromRunnable( () -> {
-            try {
-                while (!cancellationToken.isCancellationRequested()) {
-                    List<Lease> allLeases = self.leaseContainer.getAllLeases().collectList().block();
-                    List<Lease> leasesToTake = self.partitionLoadBalancingStrategy.selectLeasesToTake(allLeases);
-                    for (Lease lease : leasesToTake) {
-                        self.partitionController.addOrUpdateLease(lease).block();
-                    }
-
-                    long remainingWork = this.leaseAcquireInterval.toMillis();
-
-                    try {
-                        while (!cancellationToken.isCancellationRequested() && remainingWork > 0) {
-                            Thread.sleep(100);
-                            remainingWork -= 100;
-                        }
-                    } catch (InterruptedException ex) {
-                        // exception caught
-                        logger.warn("Partition load balancer caught an interrupted exception", ex);
+        return Flux.just(self)
+            .flatMap(value -> self.leaseContainer.getAllLeases())
+            .collectList()
+            .flatMap(allLeases -> {
+                if (cancellationToken.isCancellationRequested()) return Mono.empty();
+                List<Lease> leasesToTake = self.partitionLoadBalancingStrategy.selectLeasesToTake(allLeases);
+
+                if (cancellationToken.isCancellationRequested()) return Mono.empty();
+                return Flux.fromIterable(leasesToTake)
+                    .flatMap(lease -> {
+                        if (cancellationToken.isCancellationRequested()) return Mono.empty();
+                        return self.partitionController.addOrUpdateLease(lease);
+                    })
+                    .then();
+            })
+            .repeat(() -> {
+                if (cancellationToken.isCancellationRequested()) return false;
+
+                long remainingWork = this.leaseAcquireInterval.toMillis();
+
+                try {
+                    while (!cancellationToken.isCancellationRequested() && remainingWork > 0) {
+                        Thread.sleep(100);
```

Comment:
this is blocking call!

Reply:
We need to loop/wait before making another check to retrieve the state. Can you please recommend another way to do this in a non-blocking fashion?

Reply:
Take a look at `Flux.repeat()` and `Flux.takeUntil()` you can build a non-blocking while loop using them.

Reply:
Fixed.